### PR TITLE
libc/string: Use Byte-Shift algorithm for very long needles

### PR DESCRIPTION
**Pull request by anjiahao1**
## Testing
pass ci

### CODE REVIEW

**Comment by pkarashchenko** [comment]
So this PR only change a license header?

**Comment by anjiahao1** [comment]
> So this PR only change a license header?

it have 3 commits,the last just change license

**Comment by pkarashchenko** [comment]
> > So this PR only change a license header?
> 
> it have 3 commits,the last just change license

yeah. seems like I clicked a wrong link initially.

File: libs/libc/string/lib_strstr.c [hunk 1]
```diff
@@ -25,67 +34,198 @@
...
-      return (FAR char *)str;
+  /* i_haystack now references the (needle_len + 1)-th character. */
+
+  if (*i_needle) /* haystack is smaller than needle. */
+    {
+      return NULL;
     }
+  else if (identical)
+    {
+      return (FAR char *) haystack;
```
Comment (pkarashchenko) [inline libs/libc/string/lib_strstr.c:106]:
```suggestion
      return (FAR char *)haystack;
```

File: libs/libc/string/lib_strstr.c [hunk 2]
```diff
@@ -25,67 +34,198 @@
...
-      candidate++;
+          if (last_haystack_chars == last_needle_chars)
+            {
+              return (FAR char *)(i_haystack - needle_len);
+            }
+        }
     }
+  else /* needle_len < LONG_INT_N_BYTES */
+    {
+      mask = (((unsigned long) 1) << (needle_len * 8)) - 1;
```
Comment (pkarashchenko) [inline libs/libc/string/lib_strstr.c:209]:
```suggestion
      mask = (1lu << (needle_len * 8)) - 1;
```

**Review by pkarashchenko — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**